### PR TITLE
Fix transformers optimizations for GPT-NeoX

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description
Fix some issues found in GPT-NeoX graph fusion:
(1) GPT-NeoX uses float16 weights. The step of using onnxruntime with opt_level==1 uses CPU provider. Since most operators does not have fp16 in CPU EP, so extra Cast nodes are added to up cast to fp32.
(2) Add is shared by two LayerNormalization  children, and SkipLayerNormalization might cause invalid graph.
(3) Reshape fusion might miss since some part only check initializer but not Constant.

This PR adds a check whether model uses FP16, and output a warning when use_gpu is not True.

Note that we will add attention fusion with rotary embedding later in another pull request.

### CODE REVIEW

**Comment by keshav340** [comment]
We have to check whether weight  is ussing fp16 or not if yes  then we have to set usegpu not to true. Can i work on this issue.

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 1]
```diff
@@ -1183,3 +1184,47 @@ def add_prefix_to_names(self, prefix: str):
...
+                        return True
+
+                for node in graph.node:
+                    if node.op_type == "Cast":
+                        for attr in node.attribute:
+                            if attr.name == "to" and attr.i == TensorProto.FLOAT16:
+                                return True
+
+                    for attr in node.attribute:
+                        sub_graphs.append(attr.g)
```
Comment (yufenglee) [inline onnxruntime/python/tools/transformers/onnx_model.py:1216]: >attr.g [](http://example.com/codeflow?start=42&length=6)

check if g exists?

**Review by yufenglee — APPROVED** [review]
:shipit:

**Review by yufenglee — APPROVED** [review]

**Comment by tianleiwu** [comment]
> We have to check whether weight is ussing fp16 or not if yes then we have to set usegpu not to true. Can i work on this issue.
It is just a warning. When onnx model has fp16 weights, user could still use CPU EP for experiment.

**Pull request merged**